### ISSUE TITLE
GH Windows: ERROR: Failed to build gem native extension.

### ISSUE THREAD

**Issue by Intelligent2013** [metanorma/mnconvert#391]
Source: https://github.com/metanorma/mnconvert/actions/runs/4669609562/jobs/8268298786

```
Run gem install metanorma-cli
ERROR:  Error installing metanorma-cli:
	ERROR: Failed to build gem native extension.

    current directory: C:/hostedtoolcache/windows/Ruby/2.7.8/x64/lib/ruby/gems/2.7.0/gems/ox-2.14.15/ext/ox
C:/hostedtoolcache/windows/Ruby/2.7.8/x64/bin/ruby.exe extconf.rb
>>>>> Creating Makefile for ruby version 2.7.8 on x64-mingw32 <<<<<
checking for rb_time_timespec()... yes
checking for rb_struct_alloc_noinit()... yes
checking for rb_obj_encoding()... yes
checking for rb_ivar_foreach()... yes
checking for rb_ext_ractor_safe() in ruby.h... no
checking for pthread_mutex_init()... yes
checking for rb_enc_interned_str()... no
checking for rb_time_nano_new()... yes
checking for index()... no
checking for ruby/st.h... yes
checking for sys/uio.h... no
checking for struct tm.tm_gmtoff... no
creating Makefile
rm: cannot remove 'mkmf.log': Device or resource busy
make: [Makefile:176: clean] Error 1 (ignored)

current directory: C:/hostedtoolcache/windows/Ruby/2.7.8/x64/lib/ruby/gems/2.7.0/gems/ox-2.14.15/ext/ox
make.exe DESTDIR\= sitearchdir\=./.gem.20230411-2076-d1hhya sitelibdir\=./.gem.20230411-2076-d1hhya clean

current directory: C:/hostedtoolcache/windows/Ruby/2.7.8/x64/lib/ruby/gems/2.7.0/gems/ox-2.14.15/ext/ox
make.exe DESTDIR\= sitearchdir\=./.gem.20230411-2076-d1hhya sitelibdir\=./.gem.20230411-2076-d1hhya
generating ox-x64-mingw32.def
compiling base64.c
compiling builder.c
compiling cache.c
compiling cache8.c
compiling dump.c
compiling err.c
compiling gen_load.c
compiling hash_load.c
compiling intern.c
compiling obj_load.c
compiling ox.c
compiling parse.c
compiling sax.c
sax.c: In function 'read_element_start':
sax.c:871:21: warning: implicit declaration of function 'strndup' [-Wimplicit-function-declaration]
  871 |             ename = strndup(dr->buf.str, nlen);
      |                     ^~~~~~~
sax.c:871:21: warning: incompatible implicit declaration of built-in function 'strndup' [-Wbuiltin-declaration-mismatch]
compiling sax_as.c
compiling sax_buf.c
sax_buf.c: In function 'read_from_str':
sax_buf.c:201:5: warning: 'strncpy' specified bound depends on the length of the source argument [-Wstringop-truncation]
  201 |     strncpy(buf->tail, buf->in.str, cnt);
      |     ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
sax_buf.c:197:11: note: length computed here
  197 |     cnt = strlen(buf->in.str) + 1;
      |           ^~~~~~~~~~~~~~~~~~~
compiling sax_hint.c
compiling slotcache.c
compiling special.c
linking shared-object ox.so
C:/msys64/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/12.2.0/../../../../x86_64-w64-mingw32/bin/ld.exe: sax.o:sax.c:(.text+0x4fc1): undefined reference to `strndup'
collect2.exe: error: ld returned 1 exit status
make: *** [Makefile:[26](https://github.com/metanorma/mnconvert/actions/runs/4669609562/jobs/8268298786#step:11:27)3: ox.so] Error 1

make failed, exit code 2
...
...

Building native extensions. This could take a while...
Error: Process completed with exit code 1.
```

**Comment by ronaldtse** [comment]
This looks like a failure in building the Ox gem's native extensions.  

From https://stackoverflow.com/questions/26284110/strdup-confused-about-warnings-implicit-declaration-makes-pointer-with
> The C standard library does not have such function as `strdup`. Nevertheless this popular function is usually provided by standard library implementations as an extension. In GCC implementation this function is declared in `<string.h>`

**Comment by Intelligent2013** [comment]
From a week ago build (https://github.com/metanorma/mnconvert/actions/runs/4611668505/jobs/8151638470):
> 2023-04-04T19:47:33.7196206Z Successfully installed ox-2.14.14


From https://pullanswer.com/questions/2-14-15-cant-be-installed-on-windows:
> gem install ox -v '2.14.15' --source 'https://rubygems.org/'
> ...
> sax.c:871:21: warning: implicit declaration of function 'strndup' [-Wimplicit-function-declaration]

> gem install ox -v '2.14.14' --source 'https://rubygems.org/'
> Fetching ox-2.14.14.gem
> Temporarily enhancing PATH for MSYS/MINGW...
> Building native extensions. This could take a while...
> Successfully installed ox-2.14.14
> Parsing documentation for ox-2.14.14
> unknown encoding name ""UTF-8"" for README.md, skipping
> Installing ri documentation for ox-2.14.14
> Done installing documentation for ox after 1 seconds
> 1 gem installed

I'll try to downgrade ox gem to the version '2.14.14', as workaround solution.

**Comment by Intelligent2013** [comment]
May be no need to downgrade. Released [v2.14.16](https://github.com/ohler55/ox/releases/tag/v2.14.16) with update https://github.com/ohler55/ox/commit/4fed2b58f8b0bfbaf7f65d633cab8f04a20ad280 already.

**Comment by Intelligent2013** [comment]
No more issue:
https://github.com/metanorma/mnconvert/actions/runs/4675391351/jobs/8280467014
